crates/proof-of-sql/src/sql/postprocessing/group_by_postprocessing.rs

```diff
@@ -0,0 +1,325 @@
+                    .unwrap();
+                aggregation_expr_map.insert(key, new_col_id);
+                Expression::Column(new_col_id)
+            } else {
+                Expression::Column(*aggregation_expr_map.get(&key).unwrap())
+            }
+        }
+        Expression::Binary { op, left, right } => {
+            let left_remainder =
+                get_aggregate_and_remainder_expressions(left, aggregation_expr_map);
+            let right_remainder =
+                get_aggregate_and_remainder_expressions(right, aggregation_expr_map);
+            Expression::Binary {
+                op: *op,
+                left: Box::new(left_remainder),
+                right: Box::new(right_remainder),
+            }
+        }
+        Expression::Unary { op, expr } => {
+            let remainder = get_aggregate_and_remainder_expressions(expr, aggregation_expr_map);
+            Expression::Unary {
+                op: *op,
+                expr: Box::new(remainder),
+            }
+        }
+    }
+}
+
+/// Given an `AliasedResultExpr`, check if it is legitimate and if so grab the relevant aggregation expression
+fn check_and_get_aggregation_and_remainder(
+    expr: &AliasedResultExpr,
+    group_by_identifiers: &[Identifier],
+    aggregation_expr_map: &mut IndexMap<(AggregationOperator, Expression), Identifier>,
+) -> PostprocessingResult<AliasedResultExpr> {
+    let free_identifiers = get_free_identifiers_from_expr(&expr.expr);
+    let group_by_identifier_set = group_by_identifiers
+        .iter()
+        .cloned()
+        .collect::<IndexSet<_>>();
+    if contains_nested_aggregation(&expr.expr, false) {
+        return Err(PostprocessingError::NestedAggregationInGroupByClause(
+            format!("Nested aggregations found {:?}", expr.expr),
+        ));
+    }
+    if free_identifiers.is_subset(&group_by_identifier_set) {
+        let remainder = get_aggregate_and_remainder_expressions(&expr.expr, aggregation_expr_map);
+        Ok(AliasedResultExpr {
+            alias: expr.alias,
+            expr: Box::new(remainder),
+        })
+    } else {
+        let diff = free_identifiers
+            .difference(&group_by_identifier_set)
+            .next()
+            .unwrap();
+        Err(PostprocessingError::IdentifierNotInAggregationOperatorOrGroupByClause(*diff))
+    }
+}
+
+impl GroupByPostprocessing {
+    /// Create a new group by expression containing the group by and aggregation expressions
+    pub fn try_new(
+        by_ids: Vec<Identifier>,
+        aliased_exprs: Vec<AliasedResultExpr>,
+    ) -> PostprocessingResult<Self> {
+        let mut aggregation_expr_map: IndexMap<(AggregationOperator, Expression), Identifier> =
+            IndexMap::new();
+        // Look for aggregation expressions and check for non-aggregation expressions that contain identifiers not in the group by clause
+        let remainder_exprs: Vec<AliasedResultExpr> = aliased_exprs
+            .iter()
```

Comment:
See other ⭐ comments. Ownership could be passed around, rather than references.
```suggestion
            .into_iter()
```